helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/HelixPropertyTrimmer.java

```diff
@@ -57,38 +57,49 @@
    * @param originalProperty
    */
   protected ZNRecord doTrim(T originalProperty) {
+    ZNRecord originalZNRecord = originalProperty.getRecord();
```

Comment:
This is a high-level comment: 

Since this is causing an issue for the admin.rebalance method (which essentially just populates the mapFields in the IdealState), my immediate instinct would be that this change could be contained in IdealStateTrimmer. 

This way, we contain the scope of changes for IdealStates only. Is there any difficulty with overriding `doTrim()` in IdealStateTrimmer instead of making the change and applying it generally in HelixPropertyTrimmer?

Reply:
Good point. So this is also my first try. FYI, I also tried to extend the field type to include LIST_FIELD_KEY_ONLY. But then I find these 2 solutions are too complicated.

It is correct that putting this logic into the IdealStateTrimmer will immediately fix the problem without too much change. The complex is that in this way, it would be hard to reason the overall logic of all trimmers. Some of them care about values, the others do not. Of course, we can add comments. But gradually, it will become harder to understand and maintain.

So I go back and think about why we need these trimmers. Do we need to trim soo much information? I think we don't. The rebalancer will perform well even we detect all keys change (no value, just key's change) for all the properties. And the logic is very clean and concentrate:
1. Keys are not ignored
2. Values are ignored accordingly and this is specified by the getNonTrimmableFields() method.

Then I firstly tried to not trim all keys (including the simple fields). But then I realize it is not optimal, since the simple fields' keys are tightly bounded with the values. And finally, I have the current proposal.

Overall, the main reason for this trade-off is to make the trim logic less diverse and easier to maintain.

Reply:
BTW, the ideal solution is to separate the ZK paths for input and output. Then we can simplify or even remove the trim logic as desired.